client/apiclient.go

```diff
@@ -47,13 +50,82 @@ type TLSConfig struct {
 }
 
 // NewAPIClient gets an API client for a product.
-func NewAPIClient(product, baseURL string, headers map[string]string) *APIClient {
+func NewAPIClient(product, baseURL string, headers map[string]string, t TLSConfig) (*APIClient, error) {
+	tlsClientConfig, err := createTLSClientConfig(t)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsClientConfig,
+		},
+	}
+
 	return &APIClient{
 		Product: product,
 		BaseURL: baseURL,
 		headers: headers,
-		http:    http.DefaultClient,
+		http:    client,
+	}, nil
+}
+
+func createTLSClientConfig(t TLSConfig) (*tls.Config, error) {
+	tlsClientConfig := &tls.Config{}
+
+	// If the input TLSConfig object is default, we do not need to update tlsClientConfig.
+	if reflect.DeepEqual(t, TLSConfig{}) {
+		return tlsClientConfig, nil
 	}
+
+	if t.CACert != "" || len(t.CACertBytes) != 0 || t.CAPath != "" {
+		rootConfig := &rootcerts.Config{
+			CAFile:        t.CACert,
+			CACertificate: t.CACertBytes,
+			CAPath:        t.CAPath,
+		}
+		if err := rootcerts.ConfigureTLS(tlsClientConfig, rootConfig); err != nil {
+			return nil, err
+		}
+	}
+
+	tlsClientConfig.InsecureSkipVerify = t.Insecure
+
+	if t.TLSServerName != "" {
+		tlsClientConfig.ServerName = t.TLSServerName
+	}
+
+	var clientCert tls.Certificate
+	foundClientCert := false
+
+	switch {
+	case t.ClientCert != "" && t.ClientKey != "":
+		var err error
+		clientCert, err = tls.LoadX509KeyPair(t.ClientCert, t.ClientKey)
+		if err != nil {
+			return nil, err
+		}
+		foundClientCert = true
+	case t.ClientCert != "" || t.ClientKey != "":
+		return nil, fmt.Errorf("both client cert and client key must be provided")
+	}
+
+	if foundClientCert {
+		// We use `GetClientCertificate` here because it works with Vault, along with other products. In Vault
+		// client authentication can use a different CA than the one used for Vault's certificate. However, it
+		// will indicate that its CA should be used when sending the client certificate request. If the client
+		// certificate does not share this CA, Go will fail with a "remote error: tls: bad certificate" error.
+		// By providing an override for `GetClientCertificate`, we ensure that we send the client certificate
+		// anytime one is requested, even if the CA does not match.
+		//
+		// See GitHub issue https://github.com/hashicorp/vault/issues/2946 for more context on why Vault
```

Comment:
I'm still not sure that we neeeed this for our client, since the issue seems to be more related to vault's tls cert _auth method_ than the more general tls verification we're looking to enable, but this is such interesting sleuthing!  and if it works for all of the products just the same, then that's just delightful.  thanks for the extra commentary about it here!

Reply:
Yeah, it was hard for me to tell whether this would actually impact us or not. But, it seems to work just fine for our case, too, so I figured it would be better to avoid a potential future interop bug.